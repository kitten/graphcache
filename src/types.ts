import {
  DocumentNode,
  FragmentDefinitionNode,
  SelectionNode,
  GraphQLOutputType,
  GraphQLWrappingType,
} from 'graphql';

// Helper types
export type NullArray<T> = Array<null | T>;

// GraphQL helper types
export type SelectionSet = ReadonlyArray<SelectionNode>;
export type GraphQLFlatType = Exclude<GraphQLOutputType, GraphQLWrappingType>;
export interface Fragments {
  [fragmentName: string]: void | FragmentDefinitionNode;
}

// Scalar types are not entities as part of response data
export type Primitive = null | number | boolean | string;

export interface ScalarObject {
  __typename?: never;
  [key: string]: any;
}

export type Scalar = Primitive | ScalarObject;

export interface SystemFields {
  __typename: string;
  _id?: string | number | null;
  id?: string | number | null;
}

export type EntityField = undefined | Scalar | Scalar[];
export type DataField = Scalar | Scalar[] | Data | NullArray<Data>;

export interface DataFields {
  [fieldName: string]: DataField;
}

export interface Variables {
  [name: string]: Scalar | Scalar[] | Variables | NullArray<Variables>;
}

export type Data = SystemFields & DataFields;
export type Link<Key = string> = null | Key | NullArray<Key>;
export type ResolvedLink = Link<Data>;
export type Connection = [Variables, string];

export interface FieldInfo {
  fieldKey: string;
  fieldName: string;
  arguments: Variables | null;
}

// This is an input operation
export interface OperationRequest {
  query: DocumentNode;
  variables?: object;
}

export interface ResolveInfo {
  parentTypeName: string;
  parentKey: string;
  parentFieldKey: string;
  fieldName: string;
  fragments: Fragments;
  variables: Variables;
  partial?: boolean;
  optimistic?: boolean;
}

export interface QueryInput {
  query: string | DocumentNode;
  variables?: Variables;
}

export interface Cache {
  /** keyOfEntity() returns the key for an entity or null if it's unkeyable */
  keyOfEntity(data: Data): string | null;

  /** keyOfField() returns the key for a field */
  keyOfField(
    fieldName: string,
    args?: Variables | null | undefined
  ): string | null;

  /** resolve() retrieves the value (or link) of a field on any entity, given a partial/keyable entity or an entity key */
  resolve(
    entity: Data | string | null,
    fieldName: string,
    args?: Variables
  ): DataField;

  /** resolveFieldByKey() returns a field's value on an entity, given that field's key */
  resolveFieldByKey(entity: Data | string | null, fieldKey: string): DataField;

  /** inspectFields() retrieves all known fields for a given entity */
  inspectFields(entity: Data | string | null): FieldInfo[];

  /** invalidateQuery() invalidates all data of a given query */
  invalidateQuery(query: DocumentNode, variables?: Variables): void;

  /** updateQuery() can be used to update the data of a given query using an updater function */
  updateQuery(
    input: QueryInput,
    updater: (data: Data | null) => Data | null
  ): void;

  /** readQuery() retrieves the data for a given query */
  readQuery(input: QueryInput): Data | null;

  /** readFragment() retrieves the data for a given fragment, given a partial/keyable entity or an entity key */
  readFragment(
    fragment: DocumentNode,
    entity: string | Data,
    variables?: Variables
  ): Data | null;

  /** writeFragment() can be used to update the data of a given fragment, given an entity that is supposed to be written using the fragment */
  writeFragment(
    fragment: DocumentNode,
    data: Data,
    variables?: Variables
  ): void;
}

// Cache resolvers are user-defined to overwrite an entity field result
export type Resolver = (
  parent: Data,
  args: Variables,
  cache: Cache,
  info: ResolveInfo
) => DataField | undefined;

export interface ResolverConfig {
  [typeName: string]: {
    [fieldName: string]: Resolver;
  };
}

export type UpdateResolver = (
  result: Data,
  args: Variables,
  cache: Cache,
  info: ResolveInfo
) => void;

export type KeyGenerator = (data: Data) => null | string;

export interface UpdatesConfig {
  Mutation: {
    [fieldName: string]: UpdateResolver;
  };
  Subscription: {
    [fieldName: string]: UpdateResolver;
  };
}

export type OptimisticMutationResolver = (
  vars: Variables,
  cache: Cache,
  info: ResolveInfo
) => null | Data | NullArray<Data>;

export interface OptimisticMutationConfig {
  [mutationFieldName: string]: OptimisticMutationResolver;
}

export interface KeyingConfig {
  [typename: string]: KeyGenerator;
}

export type SerializedEntry = EntityField | Connection[] | Link;

export interface SerializedEntries {
  [key: string]: SerializedEntry;
}

export interface StorageAdapter {
  read(): Promise<SerializedEntries>;
  write(data: SerializedEntries): Promise<void>;
}

export type ErrorCode =
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 15
  | 16
  | 17
  | 18;
