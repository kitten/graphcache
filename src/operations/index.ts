export { query } from './query';
export { write, writeFragment } from './write';
export { gc } from './gc';
