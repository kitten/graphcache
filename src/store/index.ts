export {
  makeDict,
  initDataState,
  clearDataState,
  clearOptimistic,
  getCurrentDependencies,
} from './data';

export * from './keys';
export * from './store';
