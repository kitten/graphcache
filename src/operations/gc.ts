import { joinKeys } from '../helpers';
import { Store } from '../store';
import { Link } from '../types';

interface Context {
  store: Store;
  visitedLinks: Set<string>;
  visitedRecords: Set<string>;
}

export const gc = (store: Store) => {
  const visitedLinks = new Set<string>();
  const visitedRecords = new Set<string>();
  const ctx = { store, visitedLinks, visitedRecords };

  walkEntity(ctx, 'Query');

  store.records.forEach((_entity, key) => {
    if (!visitedRecords.has(key)) {
      store.remove(key);
    }
  });

  store.links.forEach((_link, key) => {
    if (!visitedLinks.has(key)) {
      store.removeLink(key);
    }
  });
};

const walkEntity = (ctx: Context, key: string) => {
  const { store, visitedRecords, visitedLinks } = ctx;
  const entity = store.find(key);

  if (entity !== null && !visitedRecords.has(key)) {
    visitedRecords.add(key);

    for (const fieldKey in entity) {
      const value = entity[fieldKey];
      if (value === undefined) {
        const childFieldKey = joinKeys(key, fieldKey);
        const link = store.readLink(childFieldKey);
        if (link !== undefined && !visitedLinks.has(childFieldKey)) {
          visitedLinks.add(childFieldKey);
          walkLink(ctx, link);
        }
      }
    }
  }
};

const walkLink = (ctx: Context, link: Link) => {
  if (Array.isArray(link)) {
    link.forEach(childLink => {
      walkLink(ctx, childLink);
    });
  } else if (link !== null) {
    walkEntity(ctx, link);
  }
};
