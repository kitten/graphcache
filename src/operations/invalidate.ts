import { FieldNode } from 'graphql';

import {
  getMainOperation,
  normalizeVariables,
  getFragments,
  getSelectionSet,
  getName,
  getFieldArguments,
} from '../ast';

import {
  EntityField,
  OperationRequest,
  Variables,
  Fragments,
  SelectionSet,
} from '../types';

import * as InMemoryData from '../store/data';
import { Store, keyOfField } from '../store';
import { SchemaPredicates } from '../ast';
import { SelectionIterator } from './shared';

interface Context {
  store: Store;
  variables: Variables;
  fragments: Fragments;
  schemaPredicates?: SchemaPredicates;
}

export const invalidate = (store: Store, request: OperationRequest) => {
  const operation = getMainOperation(request.query);

  const ctx: Context = {
    variables: normalizeVariables(operation, request.variables),
    fragments: getFragments(request.query),
    store,
    schemaPredicates: store.schemaPredicates,
  };

  invalidateSelection(
    ctx,
    ctx.store.getRootKey('query'),
    getSelectionSet(operation)
  );
};

export const invalidateSelection = (
  ctx: Context,
  entityKey: string,
  select: SelectionSet
) => {
  const isQuery = entityKey === 'Query';

  let typename: EntityField;
  if (!isQuery) {
    typename = InMemoryData.readRecord(entityKey, '__typename');
    if (typeof typename !== 'string') {
      return;
    } else {
      InMemoryData.writeRecord(entityKey, '__typename', undefined);
    }
  } else {
    typename = entityKey;
  }

  const iter = new SelectionIterator(typename, entityKey, select, ctx);

  let node: FieldNode | void;
  while ((node = iter.next()) !== undefined) {
    const fieldName = getName(node);
    const fieldKey = keyOfField(
      fieldName,
      getFieldArguments(node, ctx.variables)
    );

    if (
      process.env.NODE_ENV !== 'production' &&
      ctx.schemaPredicates &&
      typename
    ) {
      ctx.schemaPredicates.isFieldAvailableOnType(typename, fieldName);
    }

    if (node.selectionSet === undefined) {
      InMemoryData.writeRecord(entityKey, fieldKey, undefined);
    } else {
      const fieldSelect = getSelectionSet(node);
      const link = InMemoryData.readLink(entityKey, fieldKey);

      InMemoryData.writeLink(entityKey, fieldKey, undefined);
      InMemoryData.writeRecord(entityKey, fieldKey, undefined);

      if (Array.isArray(link)) {
        for (let i = 0, l = link.length; i < l; i++) {
          const childLink = link[i];
          if (childLink !== null) {
            invalidateSelection(ctx, childLink, fieldSelect);
          }
        }
      } else if (link) {
        invalidateSelection(ctx, link, fieldSelect);
      }
    }
  }
};
