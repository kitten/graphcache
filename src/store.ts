import invariant from 'invariant';
import { DocumentNode } from 'graphql';
import * as Pessimism from 'pessimism';

import {
  EntityField,
  Link,
  ResolverConfig,
  DataField,
  Variables,
  Data,
  UpdatesConfig,
  OptimisticMutationConfig,
  KeyingConfig,
} from './types';

import { keyOfEntity, joinKeys, keyOfField } from './helpers';
import { startQuery } from './operations/query';
import { writeFragment, startWrite } from './operations/write';
import { invalidate } from './operations/invalidate';

interface Ref<T> {
  current: null | T;
}

const currentDependencies: Ref<Set<string>> = { current: null };
const currentOptimisticKey: Ref<number> = { current: null };

// Resolve a ref value or throw when we're outside of a store run
const refValue = <T>(ref: Ref<T>): T => {
  invariant(
    ref.current !== null,
    'Invalid Cache call: The cache may only be accessed or mutated during' +
      'operations like write or query, or as part of its resolvers, updaters, ' +
      'or optimistic configs.'
  );

  return ref.current as T;
};

// Initialise a store run by resetting its internal state
export const initStoreState = (optimisticKey: null | number) => {
  currentDependencies.current = new Set();
  currentOptimisticKey.current = optimisticKey;
};

// Finalise a store run by clearing its internal state
export const clearStoreState = () => {
  currentDependencies.current = null;
  currentOptimisticKey.current = null;
};

export const getCurrentDependencies = () => refValue(currentDependencies);

// Add a dependency to the internal store state
export const addDependency = (dependency: string) => {
  (currentDependencies.current as Set<string>).add(dependency);
};

const mapSet = <T>(map: Pessimism.Map<T>, key: string, value: T) => {
  const optimisticKey = currentOptimisticKey.current || 0;
  return Pessimism.setOptimistic(map, key, value, optimisticKey);
};

// Used to remove a value from a Map optimistially (possible by setting it to undefined)
const mapRemove = <T>(map: Pessimism.Map<T>, key: string) => {
  const optimisticKey = currentOptimisticKey.current || 0;
  return optimisticKey
    ? Pessimism.setOptimistic(map, key, undefined, optimisticKey)
    : Pessimism.remove(map, key);
};

export class Store {
  records: Pessimism.Map<EntityField>;
  links: Pessimism.Map<Link>;

  resolvers: ResolverConfig;
  updates: UpdatesConfig;
  optimisticMutations: OptimisticMutationConfig;
  keys: KeyingConfig;

  constructor(
    resolvers?: ResolverConfig,
    updates?: Partial<UpdatesConfig>,
    optimisticMutations?: OptimisticMutationConfig,
    keys?: KeyingConfig
  ) {
    this.records = Pessimism.asMutable(Pessimism.make());
    this.links = Pessimism.asMutable(Pessimism.make());
    this.resolvers = resolvers || {};
    this.updates = {
      Mutation: (updates && updates.Mutation) || {},
      Subscription: (updates && updates.Subscription) || {},
    } as UpdatesConfig;
    this.optimisticMutations = optimisticMutations || {};
    this.keys = keys || {};
  }

  keyOfEntity(data: Data) {
    const { __typename: typename } = data;
    if (typename !== undefined && this.keys[typename] !== undefined) {
      return this.keys[typename](data);
    } else {
      return keyOfEntity(data);
    }
  }

  clearOptimistic(optimisticKey: number) {
    this.records = Pessimism.clearOptimistic(this.records, optimisticKey);
    this.links = Pessimism.clearOptimistic(this.links, optimisticKey);
  }

  getRecord(fieldKey: string): EntityField {
    return Pessimism.get(this.records, fieldKey);
  }

  removeRecord(fieldKey: string) {
    return (this.records = mapRemove(this.records, fieldKey));
  }

  writeRecord(field: EntityField, fieldKey: string) {
    return (this.records = mapSet(this.records, fieldKey, field));
  }

  getField(
    entityKey: string,
    fieldName: string,
    args?: Variables
  ): EntityField {
    const fieldKey = joinKeys(entityKey, keyOfField(fieldName, args));
    return this.getRecord(fieldKey);
  }

  writeField(
    field: EntityField,
    entityKey: string,
    fieldName: string,
    args?: Variables
  ) {
    const fieldKey = joinKeys(entityKey, keyOfField(fieldName, args));
    return (this.records = mapSet(this.records, fieldKey, field));
  }

  getLink(key: string): undefined | Link {
    return Pessimism.get(this.links, key);
  }

  removeLink(key: string) {
    return (this.links = mapRemove(this.links, key));
  }

  writeLink(link: Link, key: string) {
    return (this.links = mapSet(this.links, key, link));
  }

  resolveValueOrLink(fieldKey: string): DataField {
    const fieldValue = this.getRecord(fieldKey);
    // Undefined implies a link OR incomplete data.
    // A value will imply that we are just fetching a field like date.
    if (fieldValue !== undefined) return fieldValue;

    // This can be an array OR a string OR undefined again
    const link = this.getLink(fieldKey);
    return link ? link : null;
  }

  resolve(entity: Data | string, field: string, args?: Variables): DataField {
    if (typeof entity === 'string') {
      addDependency(entity);
      return this.resolveValueOrLink(joinKeys(entity, keyOfField(field, args)));
    } else {
      // This gives us __typename:key
      const entityKey = this.keyOfEntity(entity);
      if (entityKey === null) return null;
      addDependency(entityKey);
      return this.resolveValueOrLink(
        joinKeys(entityKey, keyOfField(field, args))
      );
    }
  }

  invalidateQuery(dataQuery: DocumentNode, variables: Variables) {
    invalidate(this, { query: dataQuery, variables });
  }

  hasField(key: string): boolean {
    return this.getRecord(key) !== undefined || this.getLink(key) !== undefined;
  }

  updateQuery(
    dataQuery: DocumentNode,
    updater: (data: Data | null) => null | Data
  ): void {
    const { data, completeness } = startQuery(this, { query: dataQuery });
    const input = completeness === 'EMPTY' ? null : data;
    const output = updater(input);
    if (output !== null) {
      startWrite(this, { query: dataQuery }, output);
    }
  }

  writeFragment(dataFragment: DocumentNode, data: Data): void {
    writeFragment(this, dataFragment, data);
  }
}
