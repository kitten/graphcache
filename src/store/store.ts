import { DocumentNode } from 'graphql';
import { createRequest } from 'urql/core';

import {
  Cache,
  FieldInfo,
  ResolverConfig,
  DataField,
  Variables,
  Data,
  QueryInput,
  UpdatesConfig,
  OptimisticMutationConfig,
  KeyingConfig,
} from '../types';

import { read, readFragment } from '../operations/query';
import { writeFragment, startWrite } from '../operations/write';
import { invalidate } from '../operations/invalidate';
import { SchemaPredicates } from '../ast';
import { keyOfField } from './keys';
import * as InMemoryData from './data';

type RootField = 'query' | 'mutation' | 'subscription';

export class Store implements Cache {
  data: InMemoryData.InMemoryData;

  resolvers: ResolverConfig;
  updates: UpdatesConfig;
  optimisticMutations: OptimisticMutationConfig;
  keys: KeyingConfig;
  schemaPredicates?: SchemaPredicates;

  rootFields: { query: string; mutation: string; subscription: string };
  rootNames: { [name: string]: RootField };

  constructor(
    schemaPredicates?: SchemaPredicates,
    resolvers?: ResolverConfig,
    updates?: Partial<UpdatesConfig>,
    optimisticMutations?: OptimisticMutationConfig,
    keys?: KeyingConfig
  ) {
    this.resolvers = resolvers || {};
    this.optimisticMutations = optimisticMutations || {};
    this.keys = keys || {};
    this.schemaPredicates = schemaPredicates;

    this.updates = {
      Mutation: (updates && updates.Mutation) || {},
      Subscription: (updates && updates.Subscription) || {},
    } as UpdatesConfig;

    if (schemaPredicates) {
      const { schema } = schemaPredicates;
      const queryType = schema.getQueryType();
      const mutationType = schema.getMutationType();
      const subscriptionType = schema.getSubscriptionType();

      const queryName = queryType ? queryType.name : 'Query';
      const mutationName = mutationType ? mutationType.name : 'Mutation';
      const subscriptionName = subscriptionType
        ? subscriptionType.name
        : 'Subscription';

      this.rootFields = {
        query: queryName,
        mutation: mutationName,
        subscription: subscriptionName,
      };

      this.rootNames = {
        [queryName]: 'query',
        [mutationName]: 'mutation',
        [subscriptionName]: 'subscription',
      };
    } else {
      this.rootFields = {
        query: 'Query',
        mutation: 'Mutation',
        subscription: 'Subscription',
      };

      this.rootNames = {
        Query: 'query',
        Mutation: 'mutation',
        Subscription: 'subscription',
      };
    }

    this.data = InMemoryData.make(this.getRootKey('query'));
  }

  gcScheduled = false;
  gc = () => {
    InMemoryData.gc(this.data);
    this.gcScheduled = false;
  };

  keyOfField = keyOfField;

  getRootKey(name: RootField) {
    return this.rootFields[name];
  }

  keyOfEntity(data: Data) {
    const { __typename: typename, id, _id } = data;
    if (!typename) {
      return null;
    } else if (this.rootNames[typename] !== undefined) {
      return typename;
    }

    let key: string | null | void;
    if (this.keys[typename]) {
      key = this.keys[typename](data);
    } else if (id !== undefined && id !== null) {
      key = `${id}`;
    } else if (_id !== undefined && _id !== null) {
      key = `${_id}`;
    }

    return key ? `${typename}:${key}` : null;
  }

  resolveFieldByKey(entity: Data | string | null, fieldKey: string): DataField {
    const entityKey =
      entity !== null && typeof entity !== 'string'
        ? this.keyOfEntity(entity)
        : entity;
    if (entityKey === null) return null;
    const fieldValue = InMemoryData.readRecord(entityKey, fieldKey);
    if (fieldValue !== undefined) return fieldValue;
    const link = InMemoryData.readLink(entityKey, fieldKey);
    return link ? link : null;
  }

  resolve(
    entity: Data | string | null,
    field: string,
    args?: Variables
  ): DataField {
    return this.resolveFieldByKey(entity, keyOfField(field, args));
  }

  invalidateQuery(query: string | DocumentNode, variables?: Variables) {
    invalidate(this, createRequest(query, variables));
  }

  inspectFields(entity: Data | string | null): FieldInfo[] {
    const entityKey =
      entity !== null && typeof entity !== 'string'
        ? this.keyOfEntity(entity)
        : entity;
    return entityKey !== null ? InMemoryData.inspectFields(entityKey) : [];
  }

  updateQuery(
    input: QueryInput,
    updater: (data: Data | null) => Data | null
  ): void {
    const request = createRequest(input.query, input.variables);
    const output = updater(this.readQuery(request as QueryInput));
    if (output !== null) {
      startWrite(this, request, output);
    }
  }

  readQuery(input: QueryInput): Data | null {
    return read(this, createRequest(input.query, input.variables)).data;
  }

  readFragment(
    dataFragment: DocumentNode,
    entity: string | Data,
    variables?: Variables
  ): Data | null {
    return readFragment(this, dataFragment, entity, variables);
  }

  writeFragment(
    dataFragment: DocumentNode,
    data: Data,
    variables?: Variables
  ): void {
    writeFragment(this, dataFragment, data, variables);
  }
}
