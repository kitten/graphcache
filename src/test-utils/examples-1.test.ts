import gql from 'graphql-tag';
import { query, write } from '../operations';
import { Store } from '../store';
import { Data } from '../types';

const Todos = gql`
  query {
    __typename
    todos {
      __typename
      id
      text
      complete
    }
  }
`;

const ToggleTodo = gql`
  mutation($id: ID!) {
    __typename
    toggleTodo(id: $id) {
      __typename
      id
      text
      complete
    }
  }
`;

const NestedClearNameTodo = gql`
  mutation($id: ID!) {
    __typename
    clearName(id: $id) {
      __typename
      todo {
        __typename
        id
        text
        complete
      }
    }
  }
`;

it('passes the "getting-started" example', () => {
  const store = new Store();
  const todosData = {
    __typename: 'Query',
    todos: [
      { id: '0', text: 'Go to the shops', complete: false, __typename: 'Todo' },
      { id: '1', text: 'Pick up the kids', complete: true, __typename: 'Todo' },
      { id: '2', text: 'Install urql', complete: false, __typename: 'Todo' },
    ],
  };

  const writeRes = write(store, { query: Todos }, todosData);

  const expectedSet = new Set(['Query.todos', 'Todo:0', 'Todo:1', 'Todo:2']);
  expect(writeRes.dependencies).toEqual(expectedSet);

  let queryRes = query(store, { query: Todos });

  expect(queryRes.data).toEqual(todosData);
  expect(queryRes.dependencies).toEqual(writeRes.dependencies);
  expect(queryRes.completeness).toBe('FULL');

  const mutatedTodo = {
    ...todosData.todos[2],
    complete: true,
  };

  const mutationRes = write(
    store,
    { query: ToggleTodo, variables: { id: '2' } },
    {
      __typename: 'Mutation',
      toggleTodo: mutatedTodo,
    }
  );

  expect(mutationRes.dependencies).toEqual(new Set(['Todo:2']));

  queryRes = query(store, { query: Todos });

  expect(queryRes.completeness).toBe('FULL');
  expect(queryRes.data).toEqual({
    ...todosData,
    todos: [...todosData.todos.slice(0, 2), mutatedTodo],
  });

  const newMutatedTodo = {
    ...mutatedTodo,
    text: '',
  };

  const newMutationRes = write(
    store,
    { query: NestedClearNameTodo, variables: { id: '2' } },
    {
      __typename: 'Mutation',
      clearName: {
        __typename: 'ClearName',
        todo: newMutatedTodo,
      },
    }
  );

  expect(newMutationRes.dependencies).toEqual(new Set(['Todo:2']));

  queryRes = query(store, { query: Todos });

  expect(queryRes.completeness).toBe('FULL');
  expect(queryRes.data).toEqual({
    ...todosData,
    todos: [...todosData.todos.slice(0, 2), newMutatedTodo],
  });
});

it('respects property-level resolvers when given', () => {
  const store = new Store({ Todo: { text: () => 'hi' } });
  const todosData = {
    __typename: 'Query',
    todos: [
      { id: '0', text: 'Go to the shops', complete: false, __typename: 'Todo' },
      { id: '1', text: 'Pick up the kids', complete: true, __typename: 'Todo' },
      { id: '2', text: 'Install urql', complete: false, __typename: 'Todo' },
    ],
  };

  const writeRes = write(store, { query: Todos }, todosData);

  const expectedSet = new Set(['Query.todos', 'Todo:0', 'Todo:1', 'Todo:2']);
  expect(writeRes.dependencies).toEqual(expectedSet);

  let queryRes = query(store, { query: Todos });

  expect(queryRes.data).toEqual({
    __typename: 'Query',
    todos: [
      { id: '0', text: 'hi', complete: false, __typename: 'Todo' },
      { id: '1', text: 'hi', complete: true, __typename: 'Todo' },
      { id: '2', text: 'hi', complete: false, __typename: 'Todo' },
    ],
  });
  expect(queryRes.dependencies).toEqual(writeRes.dependencies);
  expect(queryRes.completeness).toBe('FULL');

  const mutatedTodo = {
    ...todosData.todos[2],
    complete: true,
  };

  const mutationRes = write(
    store,
    { query: ToggleTodo, variables: { id: '2' } },
    {
      __typename: 'Mutation',
      toggleTodo: mutatedTodo,
    }
  );

  expect(mutationRes.dependencies).toEqual(new Set(['Todo:2']));

  queryRes = query(store, { query: Todos });

  expect(queryRes.completeness).toBe('FULL');
  expect(queryRes.data).toEqual({
    ...todosData,
    todos: [
      { id: '0', text: 'hi', complete: false, __typename: 'Todo' },
      { id: '1', text: 'hi', complete: true, __typename: 'Todo' },
      { id: '2', text: 'hi', complete: true, __typename: 'Todo' },
    ],
  });
});

it('Respects property-level resolvers when given', () => {
  const store = new Store(undefined, {
    Mutation: {
      toggleTodo: function toggleTodo(result, _, cache) {
        cache.updateQuery({ query: Todos }, data => {
          if (
            data &&
            data.todos &&
            result &&
            result.toggleTodo &&
            (result.toggleTodo as any).id === '1'
          ) {
            data.todos[1] = {
              id: '1',
              text: `${data.todos[1].text} (Updated)`,
              complete: (result.toggleTodo as any).complete,
              __typename: 'Todo',
            };
          } else if (data && data.todos) {
            data.todos[Number((result.toggleTodo as any).id)] = {
              ...data.todos[Number((result.toggleTodo as any).id)],
              complete: (result.toggleTodo as any).complete,
            };
          }
          return data as Data;
        });
      },
    },
  });

  const todosData = {
    __typename: 'Query',
    todos: [
      { id: '0', text: 'Go to the shops', complete: false, __typename: 'Todo' },
      {
        id: '1',
        text: 'Pick up the kids',
        complete: false,
        __typename: 'Todo',
      },
      { id: '2', text: 'Install urql', complete: false, __typename: 'Todo' },
    ],
  };

  write(store, { query: Todos }, todosData);

  write(
    store,
    { query: ToggleTodo, variables: { id: '1' } },
    {
      __typename: 'Mutation',
      toggleTodo: {
        ...todosData.todos[1],
        complete: true,
      },
    }
  );

  write(
    store,
    { query: ToggleTodo, variables: { id: '2' } },
    {
      __typename: 'Mutation',
      toggleTodo: {
        ...todosData.todos[2],
        complete: true,
      },
    }
  );

  const queryRes = query(store, { query: Todos });

  expect(queryRes.completeness).toBe('FULL');
  expect(queryRes.data).toEqual({
    ...todosData,
    todos: [
      todosData.todos[0],
      {
        id: '1',
        text: 'Pick up the kids (Updated)',
        complete: true,
        __typename: 'Todo',
      },
      { id: '2', text: 'Install urql', complete: true, __typename: 'Todo' },
    ],
  });
});
