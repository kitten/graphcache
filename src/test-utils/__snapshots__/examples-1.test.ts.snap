// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Respects property-level resolvers when given 1`] = `
Object {
  "links": Object {
    "Query.todos": Array [
      "Todo:0",
      "Todo:1",
      "Todo:2",
    ],
  },
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "todos": null,
    },
    "Todo:0": Object {
      "__typename": "Todo",
      "complete": false,
      "id": "0",
      "text": "Go to the shops",
    },
    "Todo:1": Object {
      "__typename": "Todo",
      "complete": true,
      "id": "1",
      "text": "Pick up the kids",
    },
    "Todo:2": Object {
      "__typename": "Todo",
      "complete": false,
      "id": "2",
      "text": "Install urql",
    },
  },
}
`;

exports[`Respects property-level resolvers when given 2`] = `
Object {
  "links": Object {
    "Query.todos": Array [
      "Todo:0",
      "Todo:1",
      "Todo:2",
    ],
  },
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "todos": null,
    },
    "Todo:0": Object {
      "__typename": "Todo",
      "complete": false,
      "id": "0",
      "text": "Go to the shops",
    },
    "Todo:1": Object {
      "__typename": "Todo",
      "complete": true,
      "id": "1",
      "text": "Pick up the kids",
    },
    "Todo:2": Object {
      "__typename": "Todo",
      "complete": true,
      "id": "2",
      "text": "Install urql",
    },
  },
}
`;

exports[`passes the "getting-started" example 1`] = `
Object {
  "links": Object {
    "Query.todos": Array [
      "Todo:0",
      "Todo:1",
      "Todo:2",
    ],
  },
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "todos": null,
    },
    "Todo:0": Object {
      "__typename": "Todo",
      "complete": false,
      "id": "0",
      "text": "Go to the shops",
    },
    "Todo:1": Object {
      "__typename": "Todo",
      "complete": true,
      "id": "1",
      "text": "Pick up the kids",
    },
    "Todo:2": Object {
      "__typename": "Todo",
      "complete": false,
      "id": "2",
      "text": "Install urql",
    },
  },
}
`;

exports[`passes the "getting-started" example 2`] = `
Object {
  "links": Object {
    "Query.todos": Array [
      "Todo:0",
      "Todo:1",
      "Todo:2",
    ],
  },
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "todos": null,
    },
    "Todo:0": Object {
      "__typename": "Todo",
      "complete": false,
      "id": "0",
      "text": "Go to the shops",
    },
    "Todo:1": Object {
      "__typename": "Todo",
      "complete": true,
      "id": "1",
      "text": "Pick up the kids",
    },
    "Todo:2": Object {
      "__typename": "Todo",
      "complete": true,
      "id": "2",
      "text": "Install urql",
    },
  },
}
`;
